File: pkg-py/tests/test_remote_reactome.py
"""Reactome remote mode: live existence checks via the Reactome content service."""

import pytest

import biogate
import biogate._remote as remote
from biogate import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_builds_the_content_service_query_endpoint():
    from biogate._registry import get_source

    url = remote._reactome_url(get_source("reactome"), "R-HSA-68886")
    assert url == "https://reactome.org/ContentService/data/query/R-HSA-68886"


def test_existing_stable_id_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("R-HSA-68886", source_db="reactome", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "R-HSA-68886"
    assert res.suggestion is None


def test_absent_stable_id_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biogate.check_id("R-HSA-99999999", source_db="reactome", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("r-hsa-68886", source_db="reactome", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "R-HSA-68886"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biogate.check_id("R-HSA-XYZ", source_db="reactome", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biogate.check_id("R-HSA-68886", source_db="reactome", how="remote")
    path = remote._remote_cache_path("reactome", "query", "R-HSA-68886")
    assert not path.is_file()

File: pkg-r/tests/testthat/test-remote-reactome.R
# Offline tests for reactome remote mode. Existence is checked against the
# Reactome content service; the biobouncer.remote_transport option replaces it.

.stub_reactome <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url addresses the content service query endpoint", {
  resolver <- .remote_resolvers$reactome
  url <- resolver$url(NULL, "R-HSA-68886")
  expect_identical(
    url,
    "https://reactome.org/ContentService/data/query/R-HSA-68886"
  )
})

test_that("an existing stable id is valid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_reactome(200))
  res <- check_id("R-HSA-68886", "reactome", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "R-HSA-68886")
  expect_true(is.na(res$suggestion))
})

test_that("an absent stable id is invalid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_reactome(404))
  res <- check_id("R-HSA-99999999", "reactome", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_reactome(200))
  res <- check_id("r-hsa-68886", "reactome", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "R-HSA-68886")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("R-HSA-XYZ", "reactome", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_reactome(500))
  expect_error(
    check_id("R-HSA-68886", "reactome", how = "remote"),
    class = "biobouncer_error_remote"
  )
  expect_false(file.exists(.remote_cache_path(
    "reactome",
    "query",
    "R-HSA-68886"
  )))
})
